### PR TITLE
Fixes #4996 and #4998

### LINKED ISSUE #4996: Bad handling of dummy atoms in the CIP assignment code
**Describe the bug**
Here are two demonstrations of problems:
```
In [5]: m2  = Chem.MolFromSmiles('*[C](F)(Cl)Br')

In [6]: m2.GetAtomWithIdx(1).SetChiralTag(Chem.ChiralType.CHI_TETRAHEDRAL_CW)

In [7]: Chem.FindMolChiralCenters(m2,useLegacyImplementation=False,force=True)
Out[7]: [(1, 'Tet_CW')]
```
The `Tet_CW` is coming because the CIP code doesn't add anything here. The answer should be "S" - you can demonstrate that by replacing the `*` with another light atom like `[2H]`.

And then this is just a straight-up error which should not happen:
```
In [8]: m3 = Chem.MolFromSmiles('C1CC[C@](*)2CCCC[C@H]2C1')

In [9]: Chem.FindMolChiralCenters(m3,useLegacyImplementation=False,force=True)
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-9-8e090bb78370> in <module>
----> 1 Chem.FindMolChiralCenters(m3,useLegacyImplementation=False,force=True)

~/miniconda3/envs/rdkit_blog/lib/python3.9/site-packages/rdkit/Chem/__init__.py in FindMolChiralCenters(mol, force, includeUnassigned, includeCIP, useLegacyImplementation)
    147         elif si.type == StereoType.Bond_Double:
    148           bondsToLabel.append(si.centeredOn)
--> 149       AssignCIPLabels(mol, atomsToLabel=atomsToLabel, bondsToLabel=bondsToLabel)
    150     for si in itms:
    151       if si.type == StereoType.Atom_Tetrahedral and (includeUnassigned

RuntimeError: Substituents should be topologically equivalent!
```

**Configuration (please complete the following information):**
 - RDKit version: master and current release.
 - OS: all

### PR DESCRIPTION
**Pull request by greglandrum**
Opening this to start a discussion about the fix for #4998 in particular.

What this is currently doing will clearly give the right answer if the non-kekulized bonds are not involved in distinguishing between ligands. If the bonds do need to be involved in setting chirality, it's not at all clear what the right answer should be.
We could pick bond order 1 (here) or bond order 0 (used for dative bonds).
It's worth discussing this.

### CODE REVIEW

**Comment by bp-kelley** [comment]
I know there is a solution that you dislike, but for fragments it is sometimes useful to have the bond type and bond order separately defined, otherwise I can't see any real solution to this besides the assumptions you are making.

**Comment by ricrogz** [comment]
I think this makes sense: I don't think there is a canonical answer for what priorities for dummy atoms or aromatic bonds should be, so making a decision and choosing this as a reference system makes sense to me, as long as we are consistent with it (and even better if the choice we make is also consistent with other tools).

In other words, given that there is no canonical answer to this, I'm satisfied as long as we don't get errors/exceptions in these cases and the answers we give for each case can be tracked back to one consistent reference.

**Comment by ricrogz** [comment]
> and even better if the choice we make is also consistent with other tools

Regarding this comment, I have tried this patch, and  the SMILES from #4996 in Marvin JS demo, and it seems we get opposite answers:
```
In [1]: from rdkit import Chem 
   ...: m3 = Chem.MolFromSmiles('C1CC[C@](*)2CCCC[C@H]2C1') 
   ...: Chem.FindMolChiralCenters(m3,useLegacyImplementation=False,force=True)                                                              
Out[1]: [(3, 's'), (9, 's')]
```
image

**Comment by greglandrum** [comment]
> I know there is a solution that you dislike, but for fragments it is sometimes useful to have the bond type and bond order separately defined, otherwise I can't see any real solution to this besides the assumptions you are making.

I’m actually not opposed to separating bond order and aromatic/non-aromatic (which is why the RDKit supports doing this), but that unfortunately doesn’t help if the input to the CIP algorithm is something which has aromatic bonds and can’t be kekulized

**Comment by greglandrum** [comment]
> > and even better if the choice we make is also consistent with other tools
> 
> Regarding this comment, I have tried this patch, and the SMILES from #4996 in Marvin JS demo, and it seems we get opposite answers:
> 
> ```
> In [1]: from rdkit import Chem 
>    ...: m3 = Chem.MolFromSmiles('C1CC[C@](*)2CCCC[C@H]2C1') 
>    ...: Chem.FindMolChiralCenters(m3,useLegacyImplementation=False,force=True)                                                              
> Out[1]: [(3, 's'), (9, 's')]
> ```
>

With the small amount of playing that I did in Marvin Sketch it looked like it treats the dummy as high priority. I tried replacing the dummy with an H (because it’s clear what the priority for that should be) and then assumed that the CIP label shouldn’t change if H is replaced with a dummy (which is atomic number 0 in my mind).

@ricrogz does that make sense to you?

**Comment by ricrogz** [comment]
> @ricrogz does that make sense to you?

Yes, that's what I think it's going on. My thought was whether it makes sense to try to reduce the "this other tool gives me the opposite answer" factor.

**Comment by bp-kelley** [comment]
@greglandrum A means anything but hydrogen?  While I expect a * "dummy" atom is more like R which is any atom including hydrogen.

**Comment by greglandrum** [comment]
> @greglandrum A means anything but hydrogen? While I expect a * "dummy" atom is more like R which is any atom including hydrogen.

@bp-kelley : that rendering is what marvin does when you put a "*" in a SMILES. The PR is for anything with atomic number zero.

**Comment by bp-kelley** [comment]
I was just commenting on what Marvin may be doing with the CIP rules to be consistent with the drawn A.

**Comment by bp-kelley** [comment]
Here is chemdraw with C,H,R and A

image

**Comment by greglandrum** [comment]
FWIW: BioVia Draw doesn't do CIP calculations for molecules with dummy atoms, so we can't compare that one.

So it looks like Marvin and ChemDraw are making dummy atoms really high priority. That doesn't really make sense to me, but since there's no right answer to this, I'm fine to go with that if the consensus is that it's better than having them be low priority.

**Comment by bp-kelley** [comment]
I think it probably right for "R" and wrong for single atom substitutions so they are hedging their bets.

**Comment by greglandrum** [comment]
> I think it probably right for "R" and wrong for single atom substitutions so they are hedging their bets.

Yeah, but we can't forget that they have a bunch of context in the sketcher (in the form of "R" or "A" or whatever labels) but our CIP calculation code is just getting an atom with atomic number zero, so we need a simple scheme.

**Comment by ricrogz** [comment]
> Right, I would pick the high priority as if something big was hanging off there.

 I think I'd also go with the higher priority, even if only to have some consensus with Marvin and ChemDraw. But I've realized that we are already bound to whatever `assignStereochemistry()` does (it wouldn't make sense to have contradicting priorities/labels between `assignStereochemistry()` and `assignCIPLabels()`:
 
 ```
 In [12]: from rdkit import Chem 
    ...: m = Chem.MolFromSmiles('*[C@](F)(Cl)Br') 
    ...: Chem.FindMolChiralCenters(m,useLegacyImplementation=True,force=True)                                                               
Out[12]: [(1, 'S')]
 ```
 
 Which in Marvin JS looks like:
 

So it looks like we should stay with lower priority for dummy atoms.

**Comment by greglandrum** [comment]
@ricrogz : maintaining backwards consistency in this situation where there's not an actual right answer is a good argument

File: Code/GraphMol/CIPLabeler/catch_tests.cpp [hunk 1]
```diff
@@ -467,3 +464,82 @@ TEST_CASE("para-stereochemistry", "[accurateCIP]") {
...
+  SECTION("fragment, non-unique bits") {
+    auto m = "F[C@H]([C]([NH])[CH2])c(:n):c"_smarts;
+    m->getAtomWithIdx(1)->clearProp(common_properties::_CIPCode);
+    m->updatePropertyCache();
+    CIPLabeler::assignCIPLabels(*m);
+    std::string cip;
+    CHECK(!m->getAtomWithIdx(1)->getPropIfPresent(common_properties::_CIPCode,
+                                                  cip));
+  }
+}
```
[the remark this answers is not in the document]
Comment (greglandrum) [reply Code/GraphMol/CIPLabeler/catch_tests.cpp:554]: added those tests.
Unfortunately with a rebase so the PR commit history looks more complicated than it should now... that'll go away when we squash the commits on merge.

**Review by ricrogz — APPROVED** [review]
Great!

**Pull request merged**